server/src/test/java/org/elasticsearch/action/admin/cluster/allocation/TransportDeleteDesiredBalanceActionTests.java

```diff
@@ -0,0 +1,173 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class TransportDeleteDesiredBalanceActionTests extends ESAllocationTestCase {
+
+    public void testReturnsErrorIfAllocatorIsNotDesiredBalanced() throws Exception {
+
+        @SuppressWarnings("unchecked")
+        ActionListener<ActionResponse.Empty> listener = mock(ActionListener.class);
+
+        new TransportDeleteDesiredBalanceAction(
+            mock(TransportService.class),
+            mock(ClusterService.class),
+            mock(ThreadPool.class),
+            mock(ActionFilters.class),
+            mock(IndexNameExpressionResolver.class),
+            mock(AllocationService.class),
+            mock(ShardsAllocator.class)
+        ).masterOperation(mock(Task.class), mock(DesiredBalanceRequest.class), ClusterState.EMPTY_STATE, listener);
+
+        ArgumentCaptor<ResourceNotFoundException> exceptionArgumentCaptor = ArgumentCaptor.forClass(ResourceNotFoundException.class);
+        verify(listener).onFailure(exceptionArgumentCaptor.capture());
+
+        final var exception = exceptionArgumentCaptor.getValue();
+        assertEquals("Desired balance allocator is not in use, no desired balance found", exception.getMessage());
+        assertThat(exception.status(), equalTo(RestStatus.NOT_FOUND));
+    }
+
+    public void testDeleteDesiredBalance() throws Exception {
+
+        @SuppressWarnings("unchecked")
+        ActionListener<ActionResponse.Empty> listener = mock(ActionListener.class);
+        var allocator = mock(DesiredBalanceShardsAllocator.class);
+        var allocationService = mock(AllocationService.class);
+
+        var executor = new ResetDesiredBalanceClusterExecutor(allocationService, allocator);
+        var queue = new TestMasterServiceTaskQueue<>(executor);
+
+        var clusterService = mock(ClusterService.class);
+        when(
+            clusterService.createTaskQueue(eq("reset-desired-balance"), eq(Priority.NORMAL), any(ResetDesiredBalanceClusterExecutor.class))
+        ).thenReturn(queue);
+
+        var action = new TransportDeleteDesiredBalanceAction(
+            mock(TransportService.class),
+            clusterService,
+            mock(ThreadPool.class),
+            mock(ActionFilters.class),
+            mock(IndexNameExpressionResolver.class),
+            allocationService,
+            allocator
+        );
+
+        action.masterOperation(mock(Task.class), mock(DesiredBalanceRequest.class), ClusterState.EMPTY_STATE, listener);
+        queue.processAllTasks(ClusterState.EMPTY_STATE);
+
+        verify(listener).onResponse(ActionResponse.Empty.INSTANCE);
+        verify(allocator).resetDesiredBalance();
+        verify(allocationService).reroute(any(), eq("reset-desired-balance"), any());
+    }
+
+    public static class TestMasterServiceTaskQueue<T extends ClusterStateTaskListener> implements MasterServiceTaskQueue<T> {
```

Comment:
IMO we should just use the real master service, but if you want to drop to a lower level then I recommend using `ClusterStateTaskExecutorUtils` rather than re-implementing bits of the master service yourself like this.